src/datapool/datapool_pmem.c

```diff
@@ -0,0 +1,217 @@
+        return 0;
+    }
+
+    if (pool->hdr->version != DATAPOOL_VERSION) {
+        log_info("incompatible datapool version (is: %d, expecting: %d)",
+            pool->hdr->version, DATAPOOL_SIGNATURE);
+        return 0;
+    }
+
+    if (pool->hdr->size == 0) {
+        log_error("datapool has 0 size");
+        return 0;
+    }
+
+    if (pool->hdr->size > pool->mapped_len) {
+        log_error("datapool has invalid size (is: %d, expecting: %d)",
+            pool->mapped_len, pool->hdr->size);
+        return 0;
+    }
+
+    if (pool->hdr->flags & ~DATAPOOL_VALID_FLAGS) {
+        log_error("datapool has invalid flags set");
+        return 0;
+    }
+
+    if (pool->hdr->flags & DATAPOOL_FLAG_DIRTY) {
+        log_info("datapool has a valid header but is dirty");
+        return 0;
+    }
+
+    return 1;
+}
+
+static void
+datapool_initialize(struct datapool *pool)
+{
+    log_info("initializing fresh datapool");
+
+    /* 1. clear the header from any leftovers */
+    memset(pool->hdr, 0, DATAPOOL_HEADER_LEN);
+    datapool_sync_hdr(pool);
+
+    /* 2. fill in the data */
+    pool->hdr->version = DATAPOOL_VERSION;
+    pool->hdr->size = pool->mapped_len;
+    pool->hdr->flags = 0;
+    datapool_sync_hdr(pool);
+
+    /* 3. set the signature */
+    memcpy(pool->hdr->signature, DATAPOOL_SIGNATURE, DATAPOOL_SIGNATURE_LEN);
+    datapool_sync_hdr(pool);
+}
+
+static void
+datapool_flag_set(struct datapool *pool, int flag)
+{
+    pool->hdr->flags |= flag;
+    datapool_sync_hdr(pool);
+}
+
+static void
+datapool_flag_clear(struct datapool *pool, int flag)
+{
+    pool->hdr->flags &= ~flag;
+    datapool_sync_hdr(pool);
+}
+
+/*
+ * Opens, and if necessary initializes, a datapool that resides in the given
+ * file. If no file is provided, the pool is allocated through cc_zalloc.
+ *
+ * The the datapool to retain its contents, the datapool_close() call must
+ * finish successfully.
+ */
+struct datapool *
+datapool_open(const char *path, size_t size, int *fresh)
+{
+    struct datapool *pool = cc_alloc(sizeof(*pool));
+    if (pool == NULL) {
+        log_error("unable to create allocate memory for pmem mapping");
+        goto err_alloc;
+    }
+
+    size_t map_size = size + sizeof(struct datapool_header);
+
+    if (path == NULL) { /* fallback to DRAM if pmem is not configured */
+        pool->addr = cc_zalloc(map_size);
+        pool->mapped_len = map_size;
+        pool->is_pmem = 0;
+        pool->file_backed = 0;
+    } else {
+        pool->addr = pmem_map_file(path, map_size, PMEM_FILE_CREATE, 0600,
+            &pool->mapped_len, &pool->is_pmem);
+        pool->file_backed = 1;
+    }
+
+    if (pool->addr == NULL) {
+        log_error(pmem_errormsg());
+        goto err_map;
+    }
+
+    log_info("mapped datapool %s with size %llu, is_pmem: %d",
+        path, pool->mapped_len, pool->is_pmem);
+
+    pool->hdr = pool->addr;
+    pool->user_addr = (uint8_t *)pool->addr + sizeof(struct datapool_header);
+
+    if (fresh)
```

Comment:
nit: please put brackets around conditionals, even if it's just a single statement. there are a few instances of this

Reply:
Done